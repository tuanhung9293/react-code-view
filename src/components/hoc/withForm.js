import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {clearForm, validateForm, setInitialValues} from 'state/actions/formsAction'

export function createForm(formName, FormComponent) {
  class Form extends React.Component {
    validators = {}

    componentDidMount() {
      this.setInitialValues()
      this.validate()
    }

    componentWillUnmount() {
      this.clearForm()
    }

    shouldComponentUpdate(next) {
      const {submitting, submitted, valid, dirty} = this.props
      const shouldUpdate =
        submitting !== next.submitting ||
        submitted !== next.submitted ||
        valid !== next.valid ||
        dirty !== next.dirty
      return shouldUpdate
    }

    getChildContext() {
      return {
        _form: {
          name: formName,
          registerValidator: this.registerValidator,
          unregisterValidator: this.unregisterValidator
        }
      }
    }

    render() {
      const {
        form,
        ...rest
      } = this.props
      return <FormComponent form={form} formName={formName} {...rest} handleSubmit={this.handleSubmit} />
    }

    registerValidator = (name, validator) => {
      this.validators[name] = validator
    }

    unregisterValidator = (name) => {
      delete this.validators[name]
    }

    setInitialValues = () => {
      this.props.setInitialValues()
    }

    clearForm = () => {
      this.props.clearForm()
    }

    handleSubmit = () => {
      this.props.handleSubmit(this.props.form)
    }

    validate = () => {
      const {validateForm, form: {values}} = this.props
      const errors = Object.keys(this.validators).reduce((errors, name) => {
        const value = values[name]
        const validator = this.validators[name]
        const error = validator(value)
        if (error) {
          return {
            ...errors,
            [name]: error
          }
        }

        return errors
      }, {})
      validateForm(errors)
    }
  }

  Form.childContextTypes = {
    _form: PropTypes.object.isRequired
  }

  return Form
}

function withForm(
  formName = '',
  initialValues = {},
  submitActionCreator = (values = {}) => {},
  redirectAction = () => {},
  validator = (values = {}) => {}
) {
  function mapStateToProps({forms}) {
    const form = forms[formName] || {values: initialValues, touched: {}, validationErrors: {}, meta: {}}
    const {meta, ...rest} = form
    return {
      form: rest,
      ...meta,
      dirty: Object.keys(form.values).length > 0,
      valid: Object.keys(form.validationErrors).length === 0
    }
  }

  function mapDispatchToProps(dispatch) {
    return {
      setInitialValues: () => dispatch(setInitialValues({form: formName, initialValues})),
      clearForm: () => dispatch(clearForm({form: formName})),
      handleSubmit: (form) => {
        const values = {
          ...form.initialValues,
          ...form.values
        }
        dispatch(submitActionCreator(values))
        redirectAction()
      },
      validateForm: (errors) => dispatch(validateForm({form: formName, errors}))
    }
  }

  return (Component) => connect(mapStateToProps, mapDispatchToProps)(createForm(formName, Component))
}

export default withForm