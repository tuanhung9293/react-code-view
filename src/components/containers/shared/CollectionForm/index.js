import React from 'react'
import {withRouter} from 'react-router-dom'
import PropTypes from 'prop-types'
import withForm from 'components/hoc/withForm'
import Input from 'components/ui/forms/Input'
import Textarea from 'components/ui/forms/Textarea'
import PrimaryButton from 'components/ui/buttons/PrimaryButton'
import FormField from 'components/ui/forms/FormField'
import {Columns, Column} from 'components/ui/grid'
import ModalFooter from 'components/ui/modals/ModalFooter'
import {required} from 'services/validators'
import TypeSelector from 'components/ui/forms/TypeSelector'

export function Form({dirty, valid, handleSubmit, create}) {
  const isRequired = required()

  return (
    <form onSubmit={(e) => {
      e.preventDefault()
      handleSubmit()
    }}>
      <Columns>
        <Column>
          <FormField label="Title" name="title" component={Input} />
          <FormField label="Description" name="description" component={Textarea} />
          <FormField label="Type" name="mediaType" component={TypeSelector} />
          <FormField label="Link thumbnail *" name="thumbUrl" validate={isRequired} />
          <FormField label="Link media source *" name="mediaUrl" validate={isRequired} />
        </Column>
      </Columns>
      <ModalFooter>
        <PrimaryButton outlined danger to={`${create ? '/nasa-search' : '/collection'}`}>
          Cancel
        </PrimaryButton>
        <PrimaryButton outlined type="submit" disabled={!valid || (!create && !dirty)}>
          Save
        </PrimaryButton>
      </ModalFooter>
    </form>
  )
}

function CollectionForm({name, initialValues = {}, submitAction, redirectAction, ...rest}) {
  const ConnectedForm = withForm(name, initialValues, submitAction, redirectAction)(Form)
  return <ConnectedForm {...rest} />
}

CollectionForm.propTypes = {
  name: PropTypes.string.isRequired,
  initialValues: PropTypes.object,
  submitAction: PropTypes.func.isRequired
}

export default withRouter(CollectionForm)