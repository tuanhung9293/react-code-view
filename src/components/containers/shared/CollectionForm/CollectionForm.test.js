import React from 'react'
import {shallow} from 'enzyme'
import withForm from 'components/hoc/withForm'
import configureStore from 'redux-mock-store'
const mockStore = configureStore()

import {Form} from './'
import CollectionForm from './'

const initialValues = {
  title: "title",
  description: "description"
}
const name = "addMedia"
const submitAction = () => {}
const redirectAction = () => {}

let store
beforeEach(() => {
  store = mockStore({
    forms: {
      addMedia: {
        values: {
          title: "title",
          description: "description"
        },
        validationErrors: {}
      }
    }
  })
  store.dispatch = () => {}
})

describe('renders', () => {
  it('renders CollectionForm', () => {
    expect(shallow(
      <CollectionForm
        create
        name={name}
        initialValues={initialValues}
        submitAction={submitAction}
        redirectAction={redirectAction}
      />
    )).toMatchSnapshot()
  })
  
  it('renders Form', () => {
    expect(shallow(<Form />)).toMatchSnapshot()
  })

  it('renders Form with withForm', () => {
    const FormMockComponent = withForm(name, initialValues, submitAction, redirectAction)(Form)
    const enhancedComponent = shallow(<FormMockComponent store={store} />).dive()
    expect(enhancedComponent).toMatchSnapshot()
  })
})

