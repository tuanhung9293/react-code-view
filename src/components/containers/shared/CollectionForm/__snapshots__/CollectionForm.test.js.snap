// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders renders CollectionForm 1`] = `
<Route>
  <Component />
</Route>
`;

exports[`renders renders Form 1`] = `
<form
  onSubmit={[Function]}
>
  <Columns>
    <Connect(Column)>
      <ConnectedFormFieldWithContext
        component={[Function]}
        label="Title"
        name="title"
      />
      <ConnectedFormFieldWithContext
        component={[Function]}
        label="Description"
        name="description"
      />
      <ConnectedFormFieldWithContext
        component={[Function]}
        label="Type"
        name="mediaType"
      />
      <ConnectedFormFieldWithContext
        label="Link thumbnail *"
        name="thumbUrl"
        validate={[Function]}
      />
      <ConnectedFormFieldWithContext
        label="Link media source *"
        name="mediaUrl"
        validate={[Function]}
      />
    </Connect(Column)>
  </Columns>
  <ModalFooter>
    <PrimaryButton
      danger={true}
      outlined={true}
      to="/collection"
    >
      Cancel
    </PrimaryButton>
    <PrimaryButton
      disabled={true}
      outlined={true}
      type="submit"
    >
      Save
    </PrimaryButton>
  </ModalFooter>
</form>
`;

exports[`renders renders Form with withForm 1`] = `
<Form
  clearForm={[Function]}
  dirty={true}
  form={
    Object {
      "validationErrors": Object {},
      "values": Object {
        "description": "description",
        "title": "title",
      },
    }
  }
  formName="addMedia"
  handleSubmit={[Function]}
  setInitialValues={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  valid={true}
  validateForm={[Function]}
/>
`;
